risc0/zkvm/sdk/rust/guest/src/sha.rs

```diff
@@ -12,132 +12,178 @@
+        digest_ref
+    }
 }
 
-static mut CUR_DESC: usize = 0;
+// Computes a raw digest of the given slice, and stores the digest in
+// the given pointer.  The digest memory must be uninitilaized.
+pub(crate) unsafe fn raw_digest_to(data: &[u32], digest: *mut [u32; DIGEST_WORDS]) {
+    assert_eq!(data.len() % CHUNK_SIZE, 0);
+    let type_count = data.len() / CHUNK_SIZE;
+    assert_ne!(type_count, 0);
 
-// Compute the padded size for data of size 'len' which is equal to:
-// len + 1 (terminating byte) + sizeof(uint64_t),
-// rounded up to nearest multiple of 64.
-fn padded_size(size: usize) -> usize {
-    align_up(size + 1 + mem::size_of::<u64>(), 64)
+    let desc_ptr = SHA_DESC_ZONE.alloc(1);
+
+    desc_ptr.write_volatile(SHADescriptor {
+        type_count,
+        idx: 0,
+        source: data.as_ptr() as usize,
+        digest: digest as usize,
+    });
+
+    GPIO_SHA.write_volatile(desc_ptr);
 }
 
-fn get_cur_desc() -> *mut SHADescriptor {
-    unsafe { (REGION_SHA_START as *mut SHADescriptor).add(CUR_DESC) }
+// ShaBuf represents a buffer we can use to run SHA256 on.  This way
+// we can hash using a dynamically allocated buffer, or using one of
+// the fixed memory regions.
+pub trait ShaBuf: Extend<u32> {
+    fn len(&self) -> usize;
+    fn push(&mut self, val: u32);
+    fn as_slice(&self) -> &[u32];
+    fn as_mut_slice(&mut self) -> &mut [u32];
+    fn extend_from_slice(&mut self, data: &[u32]);
+    fn resize(&mut self, new_size: usize);
 }
 
-impl SHA256 {
-    pub fn new() -> Self {
-        SHA256 {
-            storage: Vec::with_capacity(padded_size(0)),
-        }
+impl ShaBuf for Vec<u32> {
+    fn len(&self) -> usize {
+        self.len()
     }
-
-    pub fn with_capacity(capacity: usize) -> Self {
-        SHA256 {
-            storage: Vec::with_capacity(padded_size(capacity)),
-        }
+    fn push(&mut self, val: u32) {
+        self.push(val)
     }
-
-    pub fn update<T>(&mut self, data: &T) {
-        let ptr: *const T = data;
-        let len_bytes = mem::size_of::<T>();
-        self.storage.reserve(len_bytes);
-        unsafe {
-            let end = self.storage.as_mut_ptr().add(self.storage.len());
-            end.copy_from_nonoverlapping(ptr.cast(), len_bytes);
-            self.storage.set_len(self.storage.len() + len_bytes);
-        }
+    fn as_slice(&self) -> &[u32] {
+        self.as_slice()
     }
-
-    pub fn update_slice<T>(&mut self, data: &[T]) {
-        let ptr = data.as_ptr();
-        let len_bytes = data.len() * mem::size_of::<T>();
-        self.storage.reserve(len_bytes);
-        unsafe {
-            let end = self.storage.as_mut_ptr().add(self.storage.len());
-            end.copy_from_nonoverlapping(ptr.cast(), len_bytes);
-            self.storage.set_len(self.storage.len() + len_bytes);
-        }
+    fn as_mut_slice(&mut self) -> &mut [u32] {
+        self.as_mut_slice()
+    }
+    fn extend_from_slice(&mut self, data: &[u32]) {
```

Comment:
Please separate each `fn` with a newline
```suggestion
    }
    
    fn extend_from_slice(&mut self, data: &[u32]) {
```